File: services/device/src/main/java/com/example/iter/device/service/EquipmentImageCleanupService.java
package com.example.iter.device.service;

import com.example.iter.device.config.DeviceAsyncConfig;
import com.example.iter.device.storage.EquipmentImageStorage;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class EquipmentImageCleanupService {

    private final EquipmentImageStorage imageStorage;

    @Async(DeviceAsyncConfig.S3_TASK_EXECUTOR)
    public void deleteAll(List<String> objectKeys, String failureMessage) {
        objectKeys.stream()
                .filter(objectKey -> objectKey != null && !objectKey.isBlank())
                .forEach(objectKey -> deleteQuietly(objectKey, failureMessage));
    }

    private void deleteQuietly(String objectKey, String failureMessage) {
        try {
            imageStorage.delete(objectKey);
        } catch (RuntimeException exception) {
            log.error("{}: objectKey={}", failureMessage, objectKey, exception);
        }
    }
}

File: services/device/src/main/kotlin/com/example/iter/device/service/EquipmentImageCleanupService.kt
package com.example.iter.device.service

import com.example.iter.device.config.DeviceAsyncConfig
import com.example.iter.device.storage.EquipmentImageStorage
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service

@Service
class EquipmentImageCleanupService(
    private val imageStorage: EquipmentImageStorage,
) {

    @Async(DeviceAsyncConfig.S3_TASK_EXECUTOR)
    fun deleteAll(objectKeys: List<String?>, failureMessage: String) {
        objectKeys
            .filter { !it.isNullOrBlank() }
            .forEach { deleteQuietly(it, failureMessage) }
    }

    private fun deleteQuietly(objectKey: String?, failureMessage: String) {
        try {
            imageStorage.delete(objectKey)
        } catch (exception: RuntimeException) {
            log.error("{}: objectKey={}", failureMessage, objectKey, exception)
        }
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(EquipmentImageCleanupService::class.java)
    }
}
